File: discipline_ts/core/chronic/countdown.ts
import { Duration, Instant, Branded } from "../x.ts";

export const enum CountdownStatus {
  Pending,
  Running,
  Finished,
}

const BRAND = Symbol();
const STATUS = Symbol();

export type Countdown = Branded<typeof BRAND, {
  from: Instant,
  duration: Duration,
}>;

export type CountdownPending = Countdown & { 
  [STATUS]: CountdownStatus.Pending 
};

export type CountdownRunning = Countdown & { 
  [STATUS]: CountdownStatus.Running 
};

export type CountdownFinished = Countdown & { 
  [STATUS]: CountdownStatus.Finished 
};

export const construct = (from: Instant, duration: Duration): Countdown => {
  return Branded(BRAND, {
    from,
    duration,
  });
};

export const create = (from: Instant, duration: Duration): Countdown => {
  return construct(from, duration);
};

export const getFrom = (it: Countdown): Instant => {
  return it.from;
};

export const getTill = (it: Countdown): Instant => {
  return Instant.plusOrMax(it.from, it.duration);
};

export const getTotalDuration = (it: Countdown): Duration => {
  return it.duration;
};

export const setTotalDuration = (it: Countdown, duration: Duration): void => {
  it.duration = duration;
};

export const getTimeTillStartOrZero = (it: Countdown, now: Instant): Duration => {
  return Instant.tillOrZero(now, it.from);
};

export const getTimeSinceStartOrZero = (it: Countdown, now: Instant): Duration => {
  return Instant.sinceOrZero(now, it.from);
};

export const getElapsedTimeOrZero = (it: Countdown, now: Instant): Duration => {
  return Duration.min(
    getTimeSinceStartOrZero(it, now),
    it.duration
  );
};

export const getRemainingTimeOrZero = (it: Countdown, now: Instant): Duration => {
  return Duration.minusOrZero(
    getTotalDuration(it),
    getElapsedTimeOrZero(it, now)
  );
};

export const getTimeTillFinishOrZero = (it: Countdown, now: Instant): Duration => {
  return Instant.tillOrZero(now, getTill(it));
};

export const getStatus = (it: Countdown, now: Instant): CountdownStatus => {
  if (Instant.isEarilerThan(now, it.from)) {
    return CountdownStatus.Pending;
  }

  const elapsedTime = Instant.tillOrZero(it.from, now);
  if (Duration.isShorterThanOrEqualTo(elapsedTime, it.duration)) {
    return CountdownStatus.Running;
  }

  return CountdownStatus.Finished;
};

export const isPending = (it: Countdown, now: Instant): boolean => {
  return getStatus(it, now) === CountdownStatus.Pending;
};

export const isRunning = (it: Countdown, now: Instant): boolean => {
  return getStatus(it, now) === CountdownStatus.Running;
};

export const isFinished = (it: Countdown, now: Instant): boolean => {
  return getStatus(it, now) === CountdownStatus.Finished;
};

export const extendByOrSetToMax = (it: Countdown, factor: Duration): void => {
  it.duration = Duration.plusOrMax(it.duration, factor);
};

export const match = <A, B, C>(
  it: Countdown,
  now: Instant,
  cases: {
    Pending: (it: CountdownPending) => A,
    Running: (it: CountdownRunning) => B,
    Finished: (it: CountdownFinished) => C,
  }
): A | B | C => {
  switch (getStatus(it, now)) {
    case CountdownStatus.Pending: return cases.Pending(it as CountdownPending);
    case CountdownStatus.Running: return cases.Running(it as CountdownRunning);
    case CountdownStatus.Finished: return cases.Finished(it as CountdownFinished);
  }
};

export const Countdown = {
  construct,
  create,
  getFrom,
  getTill,
  getTotalDuration,
  setTotalDuration,
  getTimeTillStartOrZero,
  getTimeSinceStartOrZero,
  getElapsedTimeOrZero,
  getRemainingTimeOrZero,
  getTimeTillFinishOrZero,
  getStatus,
  isPending,
  isRunning,
  isFinished,
  extendByOrSetToMax,
  match,
};

File: discipline4/app/src/main/java/com/example/app/chronic/Countdown.kt
package com.example.app

import com.example.app.chronic.Instant
import com.example.app.chronic.Duration

/**
 * Represents the status of a countdown
 */
enum class CountdownStatus {
  Pending,   // Hasn't started yet
  Running,   // Currently in progress
  Finished   // Completed
}

/**
 * Represents a countdown from a start instant over a duration
 */
data class Countdown(
  val from: Instant,
  var duration: Duration
) {
  companion object {
    fun create(from: Instant, duration: Duration): Countdown = Countdown(from, duration)
  }
  
  fun getFrom(): Instant = from
  fun getTill(): Instant = from.plusOrMax(duration)
  fun getTotalDuration(): Duration = duration
  fun setTotalDuration(newDuration: Duration) { 
    duration = newDuration 
  }
  
  fun getTimeTillStartOrZero(now: Instant): Duration {
    return now.tillOrZero(from)
  }

  fun getTimeSinceStartOrZero(now: Instant): Duration {
    return now.sinceOrZero(from)
  }
  
  fun getElapsedTimeOrZero(now: Instant): Duration = 
    min(getTimeSinceStartOrZero(now), duration)
  
  fun getRemainingTimeOrZero(now: Instant): Duration = 
    duration.minusOrZero(getElapsedTimeOrZero(now))
  
  fun getTimeTillFinishOrZero(now: Instant): Duration = 
    now.tillOrZero(getTill())
  
  fun getStatus(now: Instant): CountdownStatus {
    return when {
      now.isEarlierThan(from) -> CountdownStatus.Pending
      getElapsedTimeOrZero(now).isShorterThanOrEqualTo(duration) -> CountdownStatus.Running
      else -> CountdownStatus.Finished
    }
  }
  
  fun isPending(now: Instant): Boolean = getStatus(now) == CountdownStatus.Pending
  fun isRunning(now: Instant): Boolean = getStatus(now) == CountdownStatus.Running
  fun isFinished(now: Instant): Boolean = getStatus(now) == CountdownStatus.Finished
  
  fun extendByOrSetToMax(factor: Duration) {
    duration = duration.plusOrMax(factor)
  }
  
  fun <R> match(
    now: Instant,
    onPending: (Countdown) -> R,
    onRunning: (Countdown) -> R,
    onFinished: (Countdown) -> R
  ): R {
    return when (getStatus(now)) {
      CountdownStatus.Pending -> onPending(this)
      CountdownStatus.Running -> onRunning(this)
      CountdownStatus.Finished -> onFinished(this)
    }
  }
}